python/pyarrow/scalar.pxi

```diff
@@ -297,42 +297,73 @@ cdef dict _DATETIME_CONVERSION_FUNCTIONS = {}
 cdef c_bool _datetime_conversion_initialized = False
 
 
+cdef _add_micros_maybe_localize(dt, micros, tzinfo):
+    import pytz
+    if tzinfo is not None:
+        if not isinstance(tzinfo, datetime.tzinfo):
+            tzinfo = string_to_tzinfo(tzinfo)
+        dt = dt.replace(microsecond=micros, tzinfo=tzinfo)
+        return tzinfo.fromutc(dt)
+    else:
+        return dt.replace(microsecond=micros)
+
+
+cdef _datetime_from_seconds(int64_t v):
+    return datetime.datetime(1970, 1, 1) + datetime.timedelta(seconds=v)
+
+
+def _nanoseconds_to_datetime_safe(v, tzinfo):
+    if v % 1000 != 0:
+        raise ValueError("Nanosecond timestamp {} is not safely convertible "
```

Comment:
Could also be a warning instead of an error? 
Or maybe add in the error message that you can access the nanoseconds value as integer with `.value` (as alternative to installing pandas)

Reply:
I think raising an exception is fine, if it causes problems for users we can revisit. I added a note that users can use ".value"